### ISSUE TITLE
get_exchange_folder takes up a significant portion (~20%) of time spent in run_algorithm

### ISSUE THREAD

**Issue by samueldg** [enigmampc/catalyst#500]
When profiling a backtest run to find some performance bottlenecks, I found that around 20% of the time `run_algorithm` takes is spent in the `get_exchange_folder` function. This seemed excessive, and I don't see any reason for this folder to change during a backtest run (correct me if I'm wrong 😅).

Here's a flame graph, with these calls highlighted:

![profile-before](https://user-images.githubusercontent.com/4542383/46972656-ffc68800-d08c-11e8-94b4-7a6e39f861eb.png)

Looking at [the source code for `get_exchange_folder`](https://github.com/enigmampc/catalyst/blob/e82bd3cedea8dc20968c0d7eba6d03e8d1ff6177/catalyst/exchange/utils/exchange_utils.py#L64-L85), it's clear that we're doing a bunch of OS calls there, which we might want to avoid in such "hot" sections of the code (executed every time we want to get a spot price!)

For that reason, I made a manual patch on my local copy of Catalyst using Python 3's [`@lru_cache` decorator](https://docs.python.org/3/library/functools.html#functools.lru_cache), so that the exchange folder is memoized:

```patch
--- catalyst/exchange/utils/exchange_utils.py   2018-10-03 13:27:29.000000000 -0400
+++ catalyst/exchange/utils/exchange_utils.py   2018-10-03 12:46:47.000000000 -0400
@@ -4,6 +4,7 @@
 import pickle
 import shutil
 from datetime import date, datetime
+from functools import lru_cache
 
 import pandas as pd
 from catalyst.assets._assets import TradingPair
@@ -61,6 +62,7 @@
     return remote_folder
 
 
+@lru_cache(maxsize=None)
 def get_exchange_folder(exchange_name, environ=None):
     """
     The root path of an exchange folder.
```

This resulted in a significant speedup, as you can see from the "after" graph:

![profile-after](https://user-images.githubusercontent.com/4542383/46972913-b88cc700-d08d-11e8-8401-172573f3b18c.png)
(notice how `get_exchange_folder` is not even visible under `get_exchange_spot_value`).

I also noticed you have your own [`weak_lru_cache`](https://github.com/enigmampc/catalyst/blob/90d9e4e2def79b49c32607ce7da349598a117021/catalyst/utils/memoize.py#L259) implemented (probably because of the Python 2 support...), which could be used instead.

However, I'm reluctant to open a PR for that, because the `get_exchange_folder` function takes an `environ` parameter, which is possibly a dict (non-hashable!), which would break the memoizing cache. On my workflow, the fix works because only `None` is ever passed as the `environ`, but I'm not familiar enough with Catalyst's codebase to know if this `environ` is ever an actual dict, which is why I'm opening this issue instead. Hopefully you guys will have more insight here, and will be able to find a fix that prevents these OS calls from being a performance bottleneck...

For the record, memoizing this call with `lru_cache`, combined with the fix in https://github.com/enigmampc/catalyst/issues/494 corresponded to a ~30-35% performance improvement.

**Comment by lenak25** [comment]
Thanks @samueldg for another great performance improvement idea.
Your findings are really encouraging in their potential to decrease the algorithm run time.

Just wanted to update that we are looking at what you proposed.

**Comment by lenak25** [comment]
@samueldg , 

Thanks again for your contribution!
The `environ` dictionary argument passed to `get_exchange_folder` was basically originated from [here](https://github.com/enigmampc/catalyst/blob/develop/catalyst/utils/paths.py#L107) : for testing purposes, a dictionary can be passed to interpret as the os environment. 
So as a result, all the functions calling the `catalyst_root` were propagating this `environ` variable .
The performance improvement you have found is significant, thus at this point, after checking all the places in which the `get_exchange_folder` is used, I've removed the use of this argument to be able to securely add the `weak_lru_cache` decorator, as you mentioned.

This will be included , together with the other improvements you have proposed, in our next release, which is a matter of days.

Thanks again!

**Comment by samueldg** [comment]
@lenak25 Thanks for following up and digging through this! Looking forward to this release 😄

**Comment by lenak25** [comment]
@samueldg , we've just released 0.5.21 including all your performance improvements.

We'll be happy if you could also share some information on your profiling procedure.

Thanks!
